internal/generate/packagemanifest.go

```diff
@@ -0,0 +1,180 @@
+)
+
+const (
+	// File extension for all PackageManifests written by Generator.
+	packageManifestFileExt = ".package.yaml"
+)
+
+var (
+	// User-facing errors.
+
+	// ErrNoVersion if no version # has been provided
+	ErrNoVersion = errors.New("version must be set")
+
+	// Internal errors.
+
+	// ErrNoOpName if the operator name has not been set
+	ErrNoOpName = genutil.InternalError("operator name must be set")
+	// ErrNoOutputDir if the directory to write the package manifest to has not been set
+	ErrNoOutputDir = genutil.InternalError("output directory must be set")
+)
+
+type PkgOptions struct {
+	// BaseDir is a directory to look for an existing base package manifest
+	// to update.
+	BaseDir string
+	// ChannelName is operator's PackageManifest channel. If a new PackageManifest is generated
+	// or ChannelName is the only channel in the generated PackageManifest,
+	// this channel will be set to the PackageManifest's default.
+	ChannelName string
+	// IsDefaultChannel determines whether ChannelName should be the default channel in the
+	// generated PackageManifest. If true, ChannelName will be the PackageManifest's default channel.
+	// Setting this field is only necessary when more than one channel exists.
+	IsDefaultChannel bool
+}
+
+// Generate configures the Generator with opts then runs it.
+func (g Generator) GeneratePackageManifest(operatorName, version, outputDir string, opts ...*PkgOptions) error {
+	if operatorName == "" {
+		return ErrNoOpName
+	} else if version == "" {
+		return ErrNoVersion
+	} else if outputDir == "" {
+		return ErrNoOutputDir
+	}
+	myOpts := &PkgOptions{}
+	if len(opts) > 0 {
+		myOpts = opts[0]
+	}
+
+	pkg, err := g.generatePackageManifest(operatorName, version, myOpts)
+	if err != nil {
+		return err
+	}
+
+	outputWriter, err := genutil.Open(outputDir, makePkgManFileName(operatorName))
+	if err != nil {
+		return err
+	}
+
+	return genutil.WriteYAML(outputWriter, pkg)
+}
+
+// generatePackageManifest takes the input and generates the populated package manifest object
+func (g *Generator) generatePackageManifest(operatorName, version string, opts *PkgOptions) (*apimanifests.PackageManifest, error) {
+	b := bases.PackageManifest{
+		PackageName: operatorName,
+	}
+	if opts.BaseDir != "" {
+		b.BasePath = filepath.Join(opts.BaseDir, makePkgManFileName(operatorName))
+	}
+	base, err := b.GetBase()
+	if err != nil {
+		return nil, fmt.Errorf("error getting PackageManifest base: %v", err)
+	}
+
+	csvName := genutil.MakeCSVName(operatorName, version)
+	if opts.ChannelName != "" {
+		setChannels(base, opts.ChannelName, csvName)
+		sortChannelsByName(base)
+		if opts.IsDefaultChannel || len(base.Channels) == 1 {
+			base.DefaultChannelName = opts.ChannelName
+		}
+	} else if len(base.Channels) == 0 {
+		setChannels(base, "alpha", csvName)
+		base.DefaultChannelName = "alpha"
+	}
+
+	if err = validatePackageManifest(base); err != nil {
```

Comment:
Have a func Generate that also validates shows hurting principles like single responsibility and cohesion.